pkg/util/quotapool/manual_time.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package quotapool
+
+import (
+	"container/heap"
+	"sort"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+)
+
+// ManualTime is a testing implementation of TimeSource.
+type ManualTime struct {
+	mu struct {
+		syncutil.Mutex
+		now    time.Time
+		timers manualTimerQueue
+	}
+}
+
+// NewManualTime constructs a new ManualTime.
+func NewManualTime(initialTime time.Time) *ManualTime {
+	mt := ManualTime{}
+	mt.mu.now = initialTime
+	return &mt
+}
+
+var _ TimeSource = (*ManualTime)(nil)
+
+// Now returns the current time.
+func (m *ManualTime) Now() time.Time {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.mu.now
+}
+
+// NewTimer constructs a new Timer.
+func (m *ManualTime) NewTimer() Timer {
+	return &manualTimer{m: m}
+}
+
+// Advance forwards the current time by the given duration.
+func (m *ManualTime) Advance(duration time.Duration) {
+	m.AdvanceTo(m.Now().Add(duration))
+}
+
+// AdvanceTo advances the current time to t. If t is earlier than the current
+// time then AdvanceTo is a no-op.
+func (m *ManualTime) AdvanceTo(t time.Time) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if !t.After(m.mu.now) {
+		return
+	}
+	m.mu.now = t
+	for len(m.mu.timers) > 0 {
+		next := m.mu.timers[0]
+		if next.at.After(m.mu.now) {
+			break
+		}
+		next.ch <- next.at
+		heap.Pop(&m.mu.timers)
+	}
+}
+
+func (m *ManualTime) add(mt *manualTimer) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !mt.at.After(m.mu.now) {
+		mt.ch <- mt.at
+	} else {
+		heap.Push(&m.mu.timers, mt)
+	}
+}
+
+func (m *ManualTime) remove(mt *manualTimer) bool {
+	// TODO(ajwerner): Consider making this a O(log N) operation instead of an
```

Comment:
For a testing implementation this is hardly going to be relevant. But if you're using it for tenancy, then perhaps yes (and you're probably removing the reference to testing on this struct in some future commit).